File: andorid/app/src/main/java/com/gaudiyakirtan/myapplication/ui/components/Tag.kt
package com.gaudiyakirtan.myapplication.ui.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.gaudiyakirtan.myapplication.ui.theme.GaurNeutral
import com.gaudiyakirtan.myapplication.ui.theme.ShyamNeutral

/**
 * Tag component for consistent badge/tag/pill styling across the application
 *
 * @param text Text to display inside the tag
 * @param variant Color variant of the tag (default, highlight, primary, black)
 * @param size Size variant of the tag (small or normal)
 * @param uppercase Whether to display the text in uppercase
 * @param onClick Optional click handler
 * @param modifier Optional modifier for additional styling
 */
@Composable
fun Tag(
    text: String,
    variant: TagVariant = TagVariant.Default,
    size: TagSize = TagSize.Normal,
    uppercase: Boolean = false,
    onClick: (() -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    // Get colors based on variant
    val backgroundColor: Color
    val textColor: Color
    
    when (variant) {
        TagVariant.Default -> {
            val neutralColor = MaterialTheme.colorScheme.tertiary
            backgroundColor = neutralColor.copy(alpha = 0.2f)
            textColor = neutralColor
        }
        TagVariant.Highlight -> {
            val highlightColor = MaterialTheme.colorScheme.surfaceVariant
            backgroundColor = highlightColor.copy(alpha = 0.8f)
            textColor = Color.White
        }
        TagVariant.Primary -> {
            val primaryColor = MaterialTheme.colorScheme.primary
            backgroundColor = primaryColor.copy(alpha = 0.2f)
            textColor = primaryColor
        }
        TagVariant.Black -> {
            backgroundColor = Color.Black.copy(alpha = 0.3f)
            textColor = Color.White
        }
    }
    
    // Get size properties based on the TagSize
    val fontSize = when (size) {
        TagSize.Small -> 10.sp
        TagSize.Normal -> 12.sp
    }
    
    val cornerRadius = when (size) {
        TagSize.Small -> 11.dp
        TagSize.Normal -> 10.dp
    }
    
    val horizontalPadding = 10.dp
    val verticalPadding = 1.dp
    
    // Apply the tag styling
    Surface(
        color = backgroundColor,
        shape = RoundedCornerShape(cornerRadius),
        onClick = onClick ?: {},
        enabled = onClick != null,
        modifier = modifier
    ) {
        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier.padding(horizontal = horizontalPadding, vertical = verticalPadding)
        ) {
            Text(
                text = if (uppercase) text.uppercase() else text,
                fontSize = fontSize,
                fontWeight = FontWeight.Medium,
                color = textColor,
                textAlign = TextAlign.Center,
                maxLines = 1
            )
        }
    }
}

enum class TagVariant {
    Default, Highlight, Primary, Black
}

enum class TagSize {
    Small, Normal
}

File: ios/gk-ios/Views/Components/Tag.swift
import SwiftUI

enum TagVariant {
    case `default`
    case highlight
    case primary
    case black
    case custom(background: Color, text: Color)
    
    func backgroundColor() -> Color {
        switch self {
        case .default:
            return Color.neutral.opacity(0.2)
        case .highlight:
            return Color.highlight.opacity(0.8)
        case .primary:
            return Color.primary.opacity(0.2)
        case .black:
            return Color.black.opacity(0.3)
        case .custom(let background, _):
            return background
        }
    }
    
    func textColor() -> Color {
        switch self {
        case .default:
            return Color.neutral
        case .highlight, .black:
            return Color.white
        case .primary:
            return Color.primary
        case .custom(_, let text):
            return text
        }
    }
}

enum TagSize {
    case small
    case normal
    case custom(fontSize: CGFloat, cornerRadius: CGFloat, horizontalPadding: CGFloat, verticalPadding: CGFloat)
    
    var fontSize: CGFloat {
        switch self {
        case .small: return 10
        case .normal: return 12
        case .custom(let fontSize, _, _, _): return fontSize
        }
    }
    
    var cornerRadius: CGFloat {
        switch self {
        case .small: return 11
        case .normal: return 10
        case .custom(_, let cornerRadius, _, _): return cornerRadius
        }
    }
    
    var horizontalPadding: CGFloat {
        switch self {
        case .small, .normal: return 10
        case .custom(_, _, let horizontalPadding, _): return horizontalPadding
        }
    }
    
    var verticalPadding: CGFloat {
        switch self {
        case .small, .normal: return 1
        case .custom(_, _, _, let verticalPadding): return verticalPadding
        }
    }
}

struct Tag: View {
    let text: String
    var variant: TagVariant = .default
    var size: TagSize = .normal
    var uppercase: Bool = false
    var customCornerRadius: CGFloat? = nil
    var action: (() -> Void)? = nil
    
    var body: some View {
        let displayText = uppercase ? text.uppercased() : text
        
        Text(displayText)
            .font(.system(size: size.fontSize, weight: .medium))
            .foregroundColor(variant.textColor())
            .padding(.horizontal, size.horizontalPadding)
            .padding(.vertical, size.verticalPadding)
            .background(variant.backgroundColor())
            .cornerRadius(customCornerRadius ?? size.cornerRadius)
            .lineLimit(1)
            .onTapGesture {
                if let action = action {
                    action()
                }
            }
    }
}

struct TagsContainer<Content: View>: View {
    let content: Content
    var spacing: CGFloat = 4
    var direction: Axis.Set = .horizontal
    var alignment: Alignment = .topLeading
    var wrap: Bool = true
    
    init(spacing: CGFloat = 4, 
         direction: Axis.Set = .horizontal, 
         alignment: Alignment = .topLeading,
         wrap: Bool = true,
         @ViewBuilder content: () -> Content) {
        self.content = content()
        self.spacing = spacing
        self.direction = direction
        self.alignment = alignment
        self.wrap = wrap
    }
    
    var body: some View {
        if direction == .horizontal {
            HStack(alignment: .center, spacing: spacing) {
                content
            }
            .fixedSize(horizontal: !wrap, vertical: true)
        } else {
            VStack(alignment: .leading, spacing: spacing) {
                content
            }
        }
    }
}

struct Tag_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 20) {
            // Default tags
            TagsContainer {
                Tag(text: "Default small", size: .small)
                Tag(text: "Default normal")
            }
            
            // Highlight tags
            TagsContainer {
                Tag(text: "Highlight small", variant: .highlight, size: .small)
                Tag(text: "Highlight normal", variant: .highlight)
            }
            
            // Primary tags
            TagsContainer {
                Tag(text: "Primary small", variant: .primary, size: .small)
                Tag(text: "Primary normal", variant: .primary)
            }
            
            // Black tags
            TagsContainer {
                Tag(text: "Black small", variant: .black, size: .small)
                Tag(text: "Black normal", variant: .black)
            }
            
            // Custom tags
            TagsContainer {
                Tag(text: "Custom", variant: .custom(background: .purple.opacity(0.2), text: .purple))
                Tag(text: "Custom Size", size: .custom(fontSize: 14, cornerRadius: 15, horizontalPadding: 15, verticalPadding: 5))
            }
            
            // Vertical container
            TagsContainer(direction: .vertical) {
                Tag(text: "Vertical 1", variant: .primary)
                Tag(text: "Vertical 2", variant: .highlight)
            }
        }
        .padding()
    }
}